eclipse/che
# Add support for path on secure endpoints

### What does this PR do?
This PR adds a notion of `endpointOrigin` (separate from the endpoint path) of "servers" which is used as the basis for determining the `/jwt/auth` endpoint to perform the authentication against.

### Screenshot/screencast of this PR
N/A

### What issues does this PR fix or reference?
#15867 #18065

### How to test this PR?
1. Create a workspace using the below specified PR
1. Run the "Start development mode" command
1. Open the `hello-greeting-endpoint`
1. Notice that the runtime token is established, the auth cookie is set and the endpoint is accessed successfully.

There is a test che-server image for this PR: `quay.io/lkrejci/che-server:issue-15867`

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### Additional Context

The devfile for the repro steps:

```yaml
metadata:
  name: secure-endpoints-reproducer
projects:
  - name: quarkus-quickstarts
    source:
      location: 'https://github.com/quarkusio/quarkus-quickstarts.git'
      type: git
      sparseCheckoutDir: getting-started
components:
  - id: redhat/quarkus-java11/latest
    type: chePlugin
  - mountSources: true
    endpoints:
      - name: hello-greeting-endpoint
        port: 8080
        attributes:
          path: /hello/greeting/che-user
          secure: 'true'
          cookiesAuthEnabled: 'true'
    memoryLimit: 4927M
    type: dockerimage
    volumes:
      - name: m2
        containerPath: /home/user/.m2
    alias: centos-quarkus-maven
    image: 'quay.io/eclipse/che-quarkus:nightly'
    env:
      - value: >-
          -XX:MaxRAMPercentage=50.0 -XX:+UseParallelGC -XX:MinHeapFreeRatio=10
          -XX:MaxHeapFreeRatio=20 -XX:GCTimeRatio=4
          -XX:AdaptiveSizePolicyWeight=90 -Dsun.zip.disableMemoryMapping=true
          -Xms20m -Djava.security.egd=file:/dev/./urandom -Duser.home=/home/user
        name: JAVA_OPTS
      - value: $(JAVA_OPTS)
        name: MAVEN_OPTS
  - mountSources: true
    endpoints:
      - name: hello-greeting-endpoint
        port: 8080
        attributes:
          path: /hello/greeting/che-user
    command:
      - tail
    args:
      - '-f'
      - /dev/null
    memoryLimit: 32M
    type: dockerimage
    alias: ubi-minimal
    image: registry.access.redhat.com/ubi8/ubi-minimal
apiVersion: 1.0.0
commands:
  - name: Package
    actions:
      - workdir: '${CHE_PROJECTS_ROOT}/quarkus-quickstarts/getting-started'
        type: exec
        command: ./mvnw package
        component: centos-quarkus-maven
  - name: Start Development mode (Hot reload + debug)
    actions:
      - workdir: '${CHE_PROJECTS_ROOT}/quarkus-quickstarts/getting-started'
        type: exec
        command: './mvnw compile quarkus:dev'
        component: centos-quarkus-maven
  - name: Package Native
    actions:
      - workdir: '${CHE_PROJECTS_ROOT}/quarkus-quickstarts/getting-started'
        type: exec
        command: ./mvnw package -Dnative -Dmaven.test.skip
        component: centos-quarkus-maven
  - name: Start Native
    actions:
      - workdir: '${CHE_PROJECTS_ROOT}/quarkus-quickstarts/getting-started/target'
        type: exec
        command: ./getting-started-1.0-SNAPSHOT-runner
        component: ubi-minimal
  - name: Attach remote debugger
    actions:
      - referenceContent: |
          {
            "version": "0.2.0",
            "configurations": [
              {
                "type": "java",
                "request": "attach",
                "name": "Attach to Remote Quarkus App",
                "hostName": "localhost",
                "port": 5005
              }
            ]
          }
        type: vscode-launch
```
